core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -380,13 +370,54 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
             if (taskState == TaskState.RUNNING) {
                 stateMachine.transitionToRunning();
             }
-            if (isFlushing()) {
-                stateMachine.transitionToFlushing();
+            // avoid extra synchronization if no new flushing or finished task was observed
+            if (newFlushingOrFinishedTaskObserved) {
```

Comment:
The `stageState` is read before a task is added to the `finishedTasks` set. I think this breaks happens-before assumptions. It is possible that `State stageState = stateMachine.getState();` is run when the state is not yet `SCHEDULED` so the `stateMachine.transitionToFinished();` is not run while at the same time the `schedulingComplete()` transitions the stage to `SCHEDULED` but doesn't see all tasks being finished.

Reply:
Yup. You are correct.